File: 19-primes/cpp/primes.cpp
#include <iostream>
#include <vector>

std::vector<bool> createSieve(int limit) {
  std::vector<bool> primes(limit + 1, true);

  primes[0] = false;
  primes[1] = false;

  int i = 0;
  for(auto item : primes) {
    if(item == true) {
      int step = i;
      int n = i * i;

      while(n < limit) {
        primes[n] = false;
        n += step;
      }
    }

    i++;
  }

  return primes;
}

int main() {
  int target = 32;

  std::vector<bool> sieve = createSieve(target);

  for(auto item : sieve) {
    std::cout << item << std::endl;
  }

  return 0;
}

File: 19-primes/py/primes.py
"""Finds pairs of prime numbers which add up to a target number"""
import sys

# ooh type annotations
def create_sieve(limit: int) -> [bool]:
    """Creates a Sieve of Eratosthenes"""

    primes = [True] * (limit + 1)
    primes[0] = primes[1] = False

    for(i, is_prime) in enumerate(primes):
        if is_prime:
            # if you want a generator, yield here
            # we need to work from both ends, so use a list
            # yield i
            for num in range(i * i, limit, i):
                primes[num] = False

    return primes

def main():
    if len(sys.argv) < 2:
        input_string = input("Enter a target number: ")
    else:
        input_string = sys.argv[1]

    target = int(input_string)
    sieve = create_sieve(target)

    results = []
    for i in range(1, int(len(sieve) / 2) + 1):
        if sieve[i] and sieve[target - i]:
            results.append((i, target - i))

    print(results)

if __name__ == '__main__':
    main()